Implement named parameters in signatures (PPC0024)

leonerd opened the pull request:
This adds a major new ability to subroutine signatures, allowing callers to pass parameters by name/value pairs rather than by position.

```perl
sub f ($x, $y, :$alpha, :$beta = undef) { ... }

f( 123, 456, alpha => 789 );
```

Originally specified in

  https://github.com/Perl/PPCs/blob/main/ppcs/ppc0024-signature-named-parameters.md

This feature is currently considered experimental.

* This set of changes requires a perldelta entry, and it is included.

leonerd committed:
Implement named parameters in signatures (PPC0024)

This adds a major new ability to subroutine signatures, allowing callers
to pass parameters by name/value pairs rather than by position.

  sub f ($x, $y, :$alpha, :$beta = undef) { ... }

  f( 123, 456, alpha => 789 );

Originally specified in

  https://github.com/Perl/PPCs/blob/main/ppcs/ppc0024-signature-named-parameters.md

This feature is currently considered experimental.

wolfsage commented on pp.c:
```diff
@@ -7934,64 +7934,168 @@ PP(pp_multiparam)
+            else {
+                // TODO: Consider collecting up all the names of unrecognised
+                // in one string
+                croak("Unrecognized named parameter '%s' to subroutine '%" SVf "'",
```
The error includes the line number, but it's the number of where the subroutine is defined! Can we output the callers line number?

wolfsage commented on pp.c:
```diff
@@ -7934,64 +7934,168 @@ PP(pp_multiparam)
+            else {
+                // TODO: Consider collecting up all the names of unrecognised
+                // in one string
+                croak("Unrecognized named parameter '%s' to subroutine '%" SVf "'",
```
If you change these croaks to Perl_croak_caller() the errors will include the line number of the caller rather than where the subroutine is defined which would be way more helpful (and in line with what signatures do without named parameters)

tonycoz replied:
`namepv` may or may not be utf8 encoded, and this message doesn't account for that.
```
croak_caller("Unrecognized named parameter '" UTF8f "' to subroutine '%" SVf "'",
                    UTF8fARG(SvUTF8(name), namepv), S_find_runcv_name());
```
If the above changes to fetch the UTF8 name this should pass `true` instead of `SvUTF8(name)`.

leonerd replied:
fixed

wolfsage commented:
Extra parameters aren't detected if they have no key value pair. Consider:
```
sub foo (:$x) {}
foo(x => 1, bar => 1); # Unrecognized named parameter 'bar' to subroutine 'main::foo'
```
vs
```
sub foo (:$x) {}
foo(x => 1, 1); # Odd name/value argument for subroutine 'main::foo'
```

... compared with normal sub sigs which gives "Too many arguments for subroutine 'main::foo'"

wolfsage commented:
This recurses into madness consuming memory:
```
sub foo (:$x, @rest) {}
foo(x => 1, "oh");
```
```
Use of uninitialized value in signature processing at sigs line 7.
Use of uninitialized value in signature processing at sigs line 7.
Use of uninitialized value in signature processing at sigs line 7.
Use of uninitialized value in signature processing at sigs line 7.
Use of uninitialized value in signature processing at sigs line 7.
Use of uninitialized value in signature processing at sigs line 7.
Use of uninitialized value in signature processing at sigs line 7.

```

wolfsage commented on pp.c:
```diff
@@ -7934,64 +7934,168 @@ PP(pp_multiparam)
+            if(!named->is_required || !SvPADSTALE(PAD_SVl(named->padix)))
+                continue;
+
+            // TODO: Consider collecting up all the names of missing
```
Yes please (and unrecognized ones above...)

leonerd replied:
I'll fix those TODOs in a follow-up commit once the initial bit is merged. That's more of a "nice-to-have" feature.

wolfsage commented on op.c:
```diff
@@ -17093,6 +17199,54 @@ Perl_subsignature_finish(pTHX)
+
+        /* For now, build the named param array in iteration order. We'll
+         * sort it at the end.
+         */
```
Is it actually sorted? If I do `sub foo (:$z, :$x) {} foo();` the error I get is for 'z' missing, but I'd expect 'x'

leonerd replied:
Ah it's currently not sorted at all. Eventually it'll be sorted by order of the hash value - so not something that's useful for end users.

tonycoz commented on op.c:
```diff
@@ -17093,6 +17199,54 @@ Perl_subsignature_finish(pTHX)
+        struct op_multiparam_named_aux *named = aux->named + namedix;
+        namedix++;
+
+        named->namepv   = savepv(namepv);
```
This is being stored in an op, shouldn't this be `savesharedpv(namepv)`?

leonerd replied:
fixed

tonycoz commented on op.c:
```diff
@@ -1227,7 +1227,17 @@ Perl_op_clear(pTHX_ OP *o)
+            if(aux->named) {
+                for(size_t namedix = 0; namedix < aux->n_named; namedix++) {
+                    struct op_multiparam_named_aux *named = aux->named + namedix;
+                    Safefree(named->namepv);
```
To match allocating `namepv` with savesharedpv() mentioned below, this should PerlMemShared_free().

leonerd replied:
fixed

wolfsage commented:
```
sub foo(:&cat) {} # A signature parameter must start with '$', '@' or '%'
```

If we detect it's a named parameter can we give a more accurate error message? The above suggests `:@cat` is legal but of course:
```
Only a scalar parameter may be named
```

tonycoz commented on pp.c:
```diff
@@ -7934,64 +7934,168 @@ PP(pp_multiparam)
-            if (UNLIKELY(SvGMAGICAL(key)))
-                key = sv_mortalcopy(key);
+            STRLEN namelen;
+            const char *namepv = SvPV(name, namelen);
```
`namepv` may or may not be utf8 encoded, but `named->namepv` always is.

This isn't a problem for ASCII names, which are all that are allowed without use utf8, but the caller could specify the name as a string and expect Latin-1 encoding (ie. the strings would be equal in a perl string sense, but the matching here would not match.)

leonerd replied:
fixed

mauke commented:
```
$ ./perl -Ilib -E 'sub f(:$x, :$x) {} say 42'
Named parameters in signatures are experimental at -e line 1.
42
```

I think declaring the same named parameter twice in one signature should be a compiler error.

mauke commented:
```
$ ./perl -Ilib -wE 'sub f($x = 1, :$y) {} say 42'
Named parameters in signatures are experimental at -e line 1.
42
```

I think this should be a compiler error as well. It doesn't make sense to have optional positional parameters if there are required named parameters in the same signature because you can never call `f` without explicitly giving a value for `$x` anyway.

davorg commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,69 @@ If there are multiple optional positional parameters and not enough
+    foo(%params);
+
+As with positional parameters, named parameters may be declared with a
+defaulting expression to provide a value to assign into the varible if the
```
"variable"

leonerd replied:
fixed

leonerd commented:
@mauke 

> I think declaring the same named parameter twice in one signature should be a compiler error.

Fixed.

> I think this should be a compiler error as well. It doesn't make sense to have optional positional parameters if there are required named parameters in the same signature because you can never call `f` without explicitly giving a value for `$x` anyway.

Fixed.

> Something odd is happening with `eval`:

Hrmmm. Something tricky, probably because the `yyerror()` function doesn't actually have croak-like behaviour but just sets an error flag and carries on anyway. I expect something somewhere isn't checking correctly in that case. I'll have a look shortly.

xenu commented on op.c:
```diff
@@ -16969,6 +17060,42 @@ Perl_subsignature_append_slurpy(pTHX_ I32 sigil, PADOFFSET padix)
+    /* Need to build OP_PARAMTEST[ other: OP_PARAMSTORE[ default-expr ] ] */
+    OP *paramstore;
+    OP *paramtest = (OP *)alloc_LOGOP(OP_PARAMTEST,
+        paramstore = newUNOP(OP_PARAMSTORE, 0, defexpr),
```
Nit: For a moment I thought that `paramstore` was declared twice. Wouldn't moving this assignment to the declaration make this code a bit more clear?

leonerd replied:
Fixed

xenu commented on op.h:
```diff
@@ -1193,12 +1193,31 @@ struct op_argcheck_aux {
+    U32         namehash;
+    PADOFFSET   padix;
+
+    int is_required : 1;
```
`is_required` should either be made `unsigned` or changed to `bool`.

`int` bitfields with no signedness specifier have implementation-defined signedness (this is unlike "normal" variables, which are guaranteed to be signed). In practice (e.g. [in GCC](https://gcc.gnu.org/onlinedocs/gcc-15.2.0/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html)) they're usually signed. Which means the only valid values for this field are -1 and 0.

I'm surprised GCC doesn't warn about this, Clang does.

leonerd replied:
Fixed

jkeenan commented:
> Last bug now fixed by [e287ee0](https://github.com/Perl/perl5/pull/23871/commits/e287ee0722fcce0381220a84808edbc773755586) - turned out to be an action block in `perly.y` leaving the llval uninitialised, which fortunately most of the time happened to be true, but in some cases was false and upset the logic.
> 
> I believe that's all the issues fixed now

I tested the pull request with Perl5::TestEachCommit.  All commits PASSed on their own.  LGTM.

rwp0 commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,77 @@ If there are multiple optional positional parameters and not enough
+passed by the caller when the subroutine is invoked.  Whereas positional
+parameters take their values from the corresponding position in the list
+passed by the caller, named parameters are matched by name-value pairs
+from the caller, where the name matches the variable name of the parameter
```
I think , (the comma) is not needed here before "where".

leonerd replied:
It probably reads better as a separate sentence even.

rwp0 commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,77 @@ If there are multiple optional positional parameters and not enough
+Named parameters in signatures are currently considered B<experimental>, and
+their behaviour may change in future releases of Perl.  Uses of named
+parameters in a signature will currently emit a warning in the
+C<experimental::signature_named_parameters> category.
```
This could instead be a link to that category ie. within the pod  of the experimental pragma (so that the users could conveniently read more if needed).

leonerd replied:
`experimental` isn't (exactly) a core module, but also its documentation doesn't have separate sections per warning category. Nor does `warnings`.

rwp0 commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,77 @@ If there are multiple optional positional parameters and not enough
+
+To invoke this function, the caller must pass pairs of names and values to
+provide values for the parameters.  Each name must match the name of one of
+the variables, omimtting its leading C<$> sigil.  The order of these values
```
"omitting", without the second M, also as previously I find the comma to be unnecessary and confusing here too.

leonerd replied:
Fixed.

rwp0 commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,77 @@ If there are multiple optional positional parameters and not enough
+To invoke this function, the caller must pass pairs of names and values to
+provide values for the parameters.  Each name must match the name of one of
+the variables, omimtting its leading C<$> sigil.  The order of these values
+is not important; both of the following lines will cause the same behaviour.
```
Expected : (colon) rather than . (comma) in the end to introduce the following lines (code samples).

leonerd replied:
Fixed.

rwp0 commented on pod/perlsub.pod:
```diff
@@ -493,6 +580,28 @@ that the caller passed no arguments:
+
+=over 2
+
+=item *
+
+zero or more mandatory positional scalar parameters
+
+=item *
+
+zero or more optional positional scalar parameters
+
```
Maybe merge into a single list item as with the following named scalar parameters using OR (ie. mandatory or optional) for consistency in the Summary.

leonerd replied:
No; the entire point here is that there's zero-or-more mandatory, followed by zero-or-more optional, in two distinct groups. They aren't all mixed together.

rwp0 reviewed (commented):
Made a review on docs' (`perlsub.pod`) spelling etc

karenetheridge commented on pod/perldiag.pod:
```diff
@@ -4132,6 +4151,16 @@ L<perlfunc/require EXPR> and L<perlfunc/do EXPR>.
+=item Missing required named parameter '%s' to subroutine '%s'
+
+(F) A subroutine using a signature that declares a non-optional named
+parameter was invoked, but the caller did not provide a value associated
```
Avoid split infinitive:

```suggestion
(F) A subroutine was invoked that uses a signature that declares a
non-optional named parameter, but the caller did not provide a value associated
```

karenetheridge commented on pod/perldiag.pod:
```diff
@@ -7466,6 +7518,17 @@ whereabouts in the regular expression the escape was discovered.
+
+(F) A subroutine using a signature that declares named parameters but has no
+slurpy parameter was invoked, and the caller provided a name that did not
+match any declared named parameter.  The caller of the subroutine is
```
```suggestion
(F) A subroutine was invoked that uses a signature that declares named parameters but has no
slurpy parameter, and the caller provided a name that did not
match any declared named parameter.  The caller of the subroutine is
```

karenetheridge commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,77 @@ If there are multiple optional positional parameters and not enough
 
-After positional parameters, additional arguments may be captured in a
-slurpy parameter.  The simplest form of this is just an array variable:
+Since Perl 5.43.3, a signature may also declare named parameters.  Like
```
```suggestion
Since Perl 5.43.5, a signature may also declare named parameters.  Like
```

karenetheridge commented on pod/perlsub.pod:
```diff
@@ -445,8 +445,77 @@ If there are multiple optional positional parameters and not enough
+variable sigil and name.  This symbol is not part of the variable's name
+within the body of the subroutine.  Named parameters may only be declared
+after any positional parameters; further positional parameters are not
+allowed after any named ones.
```
Is this a necessary restriction? It seems to me (not knowing anything about the underlying implementation) that the only restriction needs to be that there can only be one slurpy parameter (either of the positional or named type), and it needs to come last.

karenetheridge reviewed (changes requested):
Awesome!

Just a few minor tweaks to some of the doc wording.

tonycoz commented on pp.c:
```diff
@@ -7934,64 +7936,181 @@ PP(pp_multiparam)
+
+            /* namepv / namelen are always UTF-8 */
+            STRLEN namelen;
+            const char *namepv = SvPVutf8(name, namelen);
```
If the callers SV is modifiable and isn't special (it won't overwrite references or GVs for example) this will upgrade it in place, modifying the caller's SV.

If the SV isn't utf8 you could use bytes_to_utf8_free_me() or bytes_to_utf8_temp_pv() to get the UTF8 version cheaply:
```
const char *namepv = SvPV(name, namelen);
if (!SvUTF8(name))
    namepv = bytes_to_utf8_temp_pv(namepv, &namelen);
```
which saves the cost of making a new SV, and bytes_to_utf8_temp_pv() is smart about invariants (it will return `namepv` if the string contains only invariants).

The `_temp_pv()` variant uses SAVEFREEPV() to ensure the namepv is release (if allocated), alternatively you can use the `_free_me()` variant where you need to Safefree() the allocation, but the croak() makes that difficult to use here.

leonerd replied:
Ahyes, now changed and added a test to check that non-UTF8 SVs from the caller are not upgraded like that any more.

tonycoz commented on pp.c:
```diff
@@ -7934,64 +7936,183 @@ PP(pp_multiparam)
+                SV **padentry = &PAD_SVl(padix);
+                save_clearsv(padentry);
+
+                assert(TAINTING_get || !TAINT_get);
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+
+                if(!val)
+                    val = &PL_sv_undef;
```
Coverity complains about this block, since if val can be NULL, the SvTAINTED() would have crashed.

I tried to make this crash, but encountered "Odd name/value argument for subroutine" for each of my test cases, and adding an `assert(val)` here didn't trigger on make test.

So I suspect there's some redundant code here.

wolfsage replied:
```
$ BAD=x ./perl -Ilib -t -E 'use feature "signatures"; sub foo (:$x, @y) {} foo("$ENV{BAD}");'
Named parameters in signatures are experimental at -e line 1.
Segmentation fault
```

tonycoz replied:
Thanks
